subprojects/docs/src/docs/userguide/developing-plugins/kotlin_gradle_plugins.adoc

```diff
@@ -0,0 +1,334 @@
+= Writing Gradle plugins in Kotlin
+
+In addition to <<kotlin_dsl.adoc, writing build scripts>>, Gradle also supports writing plugins in https://www.jetbrains.com/opensource/kotlin/[Kotlin].
+
+There are multiple options to write your Kotlin plugins and multiple compatibility considerations. This page is a non-exhaustive list of dependencies and compiler configurations and their impacts in terms of compatibility and usability.
+
+It also describes the different development plugins to help you set up these different options.
+
+
+[[sec:dependencies]]
+== Different dependencies
+
+Like other JVM projects, Gradle plugins are jar files with dependencies. However,  unlike other JVM projects, Gradle provides fixed versions of some dependencies at runtime. Those dependencies are `compileOnly` dependencies and need to be compatible with the version of Gradle your plugin is running in.
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== The Gradle API
+
+All plugins use the https://docs.gradle.org/current/javadoc/index.html[Gradle Core API] to configure a Gradle build. The Gradle API must be added as a build compile-time dependency using the helper function `gradleApi()`
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleApi())
+}
+----
+
+TIP: `gradleApi()` must be a `compileOnly` dependency, because the correct Gradle API will be provided by Gradle at runtime.
+
+==== Version of the Gradle API Version
+
+The Gradle API version provided by `gradleApi()` is the same as the version of Gradle used to build the plugin.
+
+For example, if the plugin is built using Gradle version 7.6.1, then `gradleApi()` also be version 7.6.1. If the plugin is later built with Gradle {gradleVersion}, then the `gradleApi()` will change to be version {gradleVersion}.
+
+This implicit link affects the Gradle versions your plugin supports. It's easy to build a plugin using one version of Gradle that has functions or classes that are not present in other versions of Gradle.
+
+This might result in runtime exceptions, such as `NoSuchMethodError`, if a plugin uses the functionality of the Gradle API that is missing in the Gradle being used to run the plugin. This might happen if the plugin is run by an older Gradle version that is missing a function or class, _or_ if the plugin uses a function or class that has been removed in newer versions (which can happen when using deprecated or internal Gradle features).
```

Comment:
```suggestion
If a plugin uses the functionality of the Gradle API that is missing in the Gradle being used to run the plugin, it might result in a runtime exception, such as `NoSuchMethodError`.
This can happen if the plugin is run by an older Gradle version that is missing a function or class.
It can also happen if the plugin uses a function or class that has been removed in newer versions (which can happen when using deprecated or internal Gradle features).
```